src/pkcs11/framework-pkcs15.c

```diff
@@ -3879,8 +3879,9 @@ pkcs15_pubkey_get_attribute(struct sc_pkcs11_session *session, void *object, CK_
 		break;
 	case CKA_TOKEN:
 	case CKA_SENSITIVE:
```

Comment:
PKCS#11 standard says that CKA_SENSITIVE is the attribute of either a private or a secret key. Take a look at the references below.

For example, [http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cs01/pkcs11-curr-v2.40-cs01.pdf] clearly states that CKA_SENSITIVE is an attribute of private keys (and secret keys),  **and omits it** from the public key templates.

Other references may be helpful too:
www.cryptsoft.com/pkcs11doc/STANDARD/pkcs-11v2-20.pdf
https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=3&cad=rja&uact=8&ved=0ahUKEwi3x9Xh7u7JAhVDbSYKHebtCLgQFggpMAI&url=https%3A%2F%2Fwww.oasis-open.org%2Fcommittees%2Fdownload.php%2F49502%2Fpkcs11-cka-public-key-info-r4.docx&usg=AFQjCNGks6-yexjogHCRZNLsUG0TbFapuw&sig2=us78EEjV4r5iqsTNhS4wCQ
https://books.google.com/books?id=tEO6BQAAQBAJ&pg=PA618&lpg=PA618&dq=pkcs11+meaning+of+key+attribute+CKA_SENSITIVE&source=bl&ots=THU1LvK1se&sig=ttcL8bHtaptxw2BC7TBx4Txn7Xg&hl=en&sa=X&ved=0ahUKEwi3x9Xh7u7JAhVDbSYKHebtCLgQ6AEIMDAD#v=onepage&q=pkcs11%20meaning%20of%20key%20attribute%20CKA_SENSITIVE&f=false

It might be interesting to learn who wrote the OpenSC code the way it is, and why - but as you pointed out yourself, the standard should take precedence.
```
@git blame src/pkcs11/framework-pkcs11.c
..........
6e6d7177 (fabled               2002-01-16 22:49:03 +0000 3142)  case CKA_TOKEN:
6e6d7177 (fabled               2002-01-16 22:49:03 +0000 3143)          check_attrib
ute_buffer(attr, sizeof(CK_BBOOL));
6e6d7177 (fabled               2002-01-16 22:49:03 +0000 3144)          *(CK_BBOOL*)attr->pValue = TRUE;
.............
6e6d7177 (fabled               2002-01-16 22:49:03 +0000 3386)  case CKA_CLASS:
6e6d7177 (fabled               2002-01-16 22:49:03 +0000 3387)          check_attrib
ute_buffer(attr, sizeof(CK_OBJECT_CLASS));
6e6d7177 (fabled               2002-01-16 22:49:03 +0000 3388)          *(CK_OBJECT_
CLASS*)attr->pValue = CKO_PRIVATE_KEY;
54f87942 (nils                 2005-02-03 22:44:19 +0000 3389)          break;
6e6d7177 (fabled               2002-01-16 22:49:03 +0000 3390)  case CKA_TOKEN:
4915eaa5 (Thomas Calderon      2014-12-03 17:59:04 +0100 3391)          check_attrib
ute_buffer(attr, sizeof(CK_BBOOL));
4915eaa5 (Thomas Calderon      2014-12-03 17:59:04 +0100 3392)          *(CK_BBOOL*)attr->pValue = TRUE;
4915eaa5 (Thomas Calderon      2014-12-03 17:59:04 +0100 3393)          break;
.................
252323e0 (okir                 2002-03-12 14:36:40 +0000 3880)  case CKA_TOKEN:
252323e0 (okir                 2002-03-12 14:36:40 +0000 3881)  case CKA_SENSITIVE:
252323e0 (okir                 2002-03-12 14:36:40 +0000 3882)          check_attrib
ute_buffer(attr, sizeof(CK_BBOOL));
00000000 (Not Committed Yet    2015-12-22 01:59:09 -0500 3883)          *(CK_BBOOL*)
attr->pValue = FALSE;
```

But according to this https://github.com/OpenSC/OpenSC/commit/4915eaa5, `case CKA_SENSITIVE:` should've been removed from that place by @viktorTarasov on 2014-12-03. His code makes it clear that this attribute makes sense **only** for private keys...